### PR TITLE
stage_executor: reset platform in systemcontext for every stage.

### PR DESCRIPTION
**Pull request by flouthoc**
On processing of every stage platform spec in systemcontext must be correctly reset.

[NO NEW TESTS NEEDED]

Closes: https://github.com/containers/buildah/issues/5968

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
stage_executor: reset platform in systemcontext for every stage.
```

### CODE REVIEW

**Comment by flouthoc** [comment]
Just wondering how to test this in CI since `binfmt` will correctly run every image type.

**Comment by flouthoc** [comment]
Temp adding `No New Tests` to see if CI passes, will remove the label when I add test.

**Comment by nalind** [comment]
The "header-builtin" test does something similar, so I'd suggest
```Dockerfile
ARG FOREIGNARCH
FROM --platform=linux/$FOREIGNARCH busybox AS foreign
FROM busybox
COPY --from=foreign /bin/busybox /bin/busybox.foreign
RUN arch
RUN ls -l /bin/busybox /bin/busybox.foreign
RUN ! cmp /bin/busybox /bin/busybox.foreign
```
Define the `FOREIGNARCH` build arg by checking if `$(go env GOARCH)` is `amd64` or `arm64` or `ppc64le` or `s390x`, having preselected one or them as a default, and choosing a different one if that default happens to match the local architecture.

File: imagebuildah/stage_executor.go [hunk 1]
```diff
@@ -59,6 +59,7 @@ import (
 // name to the image that it produces.
 type StageExecutor struct {
 	ctx                   context.Context
+	systemContext         types.SystemContext
```
Comment (nalind) [inline imagebuildah/stage_executor.go:62]: `types.SystemContext` values in structs are customarily pointers.  This should be a pointer.

File: imagebuildah/stage_executor.go [hunk 2]
```diff
@@ -952,19 +953,21 @@ func (s *StageExecutor) prepare(ctx context.Context, from string, initializeIBCo
...
 
-	builderSystemContext := s.executor.systemContext
-	// get platform string from stage
-	if stage.Builder.Platform != "" {
-		os, arch, variant, err := parse.Platform(stage.Builder.Platform)
+	builderSystemContext := s.systemContext
+	// In a multi-stage build where `FROM --platform=<>` was used then we must
+	// reset context for new stages so that new stages don't inherit unexpected
+	// `--platform` from prior stages.
+	if stage.Builder.Platform != "" || (stage.Position != 0 && builderSystemContext.ArchitectureChoice == "" && builderSystemContext.VariantChoice == "" && builderSystemContext.OSChoice == "") {
```
Comment (nalind) [inline imagebuildah/stage_executor.go:960]: Would this be simpler if the `builderSystemContext` had its `...Choice` fields cleared as part of initializing it?

File: tests/bud.bats [hunk 3]
```diff
@@ -6486,6 +6486,15 @@ _EOF
   done
 }
 
+@test "build must reset platform for stages if needed" {
+  otherarch="arm64"
+  # just make sure that other arch is not equivalent to host arch
+  if [[ "$otherarch" == "$myarch" ]]; then
```
Comment (nalind) [inline tests/bud.bats:6492]: `$myarch` has not been defined at this point, so the test is always false.

File: tests/bud.bats [hunk 4]
```diff
@@ -6486,6 +6486,15 @@ _EOF
...
+    otherarch="amd64"
+  fi
+  run_buildah build $WITH_POLICY_JSON --build-arg FOREIGNARCH=$otherarch -f $BUDFILES/multiarch/Containerfile.reset-platform $BUDFILES/multiarch
```
Comment (nalind) [inline tests/bud.bats:6495]: This should also test a case where the the TARGETPLATFORM build-arg is set, so that we don't regress on that.

File: imagebuildah/stage_executor.go (hunk 2 above)
Comment (flouthoc) [reply imagebuildah/stage_executor.go:960]: I don't think we can do that since in certain cases the fields are set beforehand for example when it's build with `--platform` or `--all-platforms` so clearing `...Choice` would reset values in such case.

File: tests/bud.bats (hunk 3 above)
Comment (flouthoc) [reply tests/bud.bats:6494]: Added.

File: tests/bud.bats (hunk 4 above)
Comment (flouthoc) [reply tests/bud.bats:6497]: I added a test which sets TARGETPLATFORM using `--platform`, should that cover what you meant by this comment ?

Comment (nalind) [reply tests/bud.bats:6497]: I mean literally setting `--build-arg TARGETPLATFORM=`, which has been known to be done instead of using the `--platform` flag.  This had to be fixed before in https://github.com/containers/buildah/pull/5731.

File: imagebuildah/stage_executor.go (hunk 2 above)
Comment (nalind) [reply imagebuildah/stage_executor.go:966]: I think there's a weird corner case in here when jobs > 1, skipUnusedStages is false, and one or more later stages are using the same base image as the first stage.

**Comment by flouthoc** [comment]
I have to get back to @nalind 's comment since I was working on infra needed for this PR https://github.com/containers/common/pull/2339

**Comment by flouthoc** [comment]
@nalind Could you PTAL again.

File: imagebuildah/stage_executor.go (hunk 2 above)
Comment (nalind) [reply imagebuildah/stage_executor.go:974]: Is the intention here that the first stage should default to being built for the default platform when `--platform` is not specified?  It's hard to tell.

Comment (flouthoc) [reply imagebuildah/stage_executor.go:974]: I forgot to push a change, apologies. Let me push it again.

Comment (nalind) [reply imagebuildah/stage_executor.go]: Anything that happens when `jobs > 1` can also happen when there are multiple builds running, as podman allows.  Given that, I don't think this recent addition actually fixes that third case.
There are also two cases where `len(s.stages) > 1` is evaluated, and that makes this line longer and, for me, that much harder to wrap my head around.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: > There are also two cases where len(s.stages) > 1 is evaluated, and that makes this line longer and, for me, that much harder to wrap my head around.

I fixed it.

> Anything that happens when jobs > 1 can also happen when there are multiple builds running, as podman allows. Given that, I don't think this recent addition actually fixes that third case.

Yes patch in this PR only covers the case when single `buildah` invoked and multiple jobs are sharing same context variables. I suspect there could be more cases in `pull` then parallel invocation of multiple buildah can also step into some new issues but to be honest I did not test that.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Okay I two experiments one with `--skip-unused-stages=false` and one without with following containerfile

```Dockerfile
FROM --platform=linux/riscv64 quay.io/prometheus/busybox AS risc
FROM --platform=linux/amd64 quay.io/prometheus/busybox AS amd
FROM --platform=linux/arm64 quay.io/prometheus/busybox AS foreign
FROM quay.io/prometheus/busybox
COPY --from=foreign /bin/busybox /bin/busybox.foreign
COPY --from=amd /bin/busybox /bin/busybox.amd
RUN arch
RUN ls -l /bin/busybox /bin/busybox.foreign
RUN ! cmp /bin/busybox /bin/busybox.foreign
RUN cmp /bin/busybox /bin/busybox.amd
```

Experiment One
```console
for i in `seq 1 25`; do ./buildah build --iidfile id.$i . & done; wait; ls -al
```

Experiment Two
```console
for i in `seq 1 25`; do ./buildah build --jobs=2 --skip-unused-stages=false --iidfile id.$i . & done; wait; ls -al
```

I ran both experiments a few times I did not see any failure.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: @nalind Should I open a different github issue (to track) to find verifying to check if `parallel` race exists or not. As I able not able to find any issues till now but wondering if this PR can get blocked because of that.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: I created an issue to track this thing separately https://github.com/containers/buildah/issues/6104 so it's easier to move with this PR.

File: imagebuildah/stage_executor.go [hunk 5]
```diff
@@ -966,19 +967,21 @@ func (s *StageExecutor) prepare(ctx context.Context, from string, initializeIBCo
...
-			builderSystemContext.ArchitectureChoice = arch
-			builderSystemContext.VariantChoice = variant
+		if imageArch != "" || imageVariant != "" {
+			builderSystemContext.ArchitectureChoice = imageArch
+			builderSystemContext.VariantChoice = imageVariant
 		}
-		if os != "" {
-			builderSystemContext.OSChoice = os
+		if imageOS != "" {
+			builderSystemContext.OSChoice = imageOS
```
Comment (nalind) [inline imagebuildah/stage_executor.go:984]: Why are we creating `builderSystemContext`, which is a copy of the pointer value in `s.systemContext`, and modifying it through that pointer?  It has the same effect as modifying `s.systemContext` directly, but is harder to follow.  Is it intended to be a one-off copy that we just pass in the `BuilderOptions`?

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Removed. This is not required since we are already copying context for each stage in `executor`.

File: tests/bud.bats [hunk 6]
```diff
@@ -6573,6 +6573,21 @@ _EOF
...
+  run_buildah build $WITH_POLICY_JSON --build-arg TARGETPLATFORM=linux/$myarch --build-arg FOREIGNARCH=$otherarch -f $BUDFILES/multiarch/Containerfile.reset-platform $BUDFILES/multiarch
+  # Following line here attempts to cover edge case discussed here https://github.com/containers/buildah/pull/5971/files#r1970277844 but it's hard to reproduce this in a certain
+  # way since it can happen only when unused stage is picked in a particular order, following line ensures if this thing breaks then we luckily catch it in our CI.
+  run_buildah build $WITH_POLICY_JSON --jobs=2 --skip-unused-stages=true --build-arg FOREIGNARCH=$otherarch -f $BUDFILES/multiarch/Containerfile.unused-stage $BUDFILES/multiarch
```
Comment (nalind) [inline tests/bud.bats:6588]: What is this `run_buildah` invocation testing, exactly?  The image that's produced isn't examined, and the `build` run's success only depends on the base image existing.

File: tests/bud.bats [hunk 7]
```diff
@@ -6573,6 +6573,21 @@ _EOF
...
+  run_buildah info --format '{{.host.arch}}'
+  myarch="$output"
+  otherarch="arm64"
+  # just make sure that other arch is not equivalent to host arch
+  if [[ "$otherarch" == "$myarch" ]]; then
+    otherarch="amd64"
+  fi
+  run_buildah build $WITH_POLICY_JSON --build-arg FOREIGNARCH=$otherarch -f $BUDFILES/multiarch/Containerfile.reset-platform $BUDFILES/multiarch
+  run_buildah build $WITH_POLICY_JSON --build-arg TARGETPLATFORM=linux/$myarch --build-arg FOREIGNARCH=$otherarch -f $BUDFILES/multiarch/Containerfile.reset-platform $BUDFILES/multiarch
+  # Following line here attempts to cover edge case discussed here https://github.com/containers/buildah/pull/5971/files#r1970277844 but it's hard to reproduce this in a certain
```
Comment (nalind) [inline tests/bud.bats:6586]: This URL no longer turns up a discussion.

File: imagebuildah/stage_executor.go [hunk 8]
```diff
@@ -966,19 +967,20 @@ func (s *StageExecutor) prepare(ctx context.Context, from string, initializeIBCo
...
 	}
 
-	builderSystemContext := s.executor.systemContext
-	// get platform string from stage
-	if stage.Builder.Platform != "" {
-		os, arch, variant, err := parse.Platform(stage.Builder.Platform)
+	// In a multi-stage build where `FROM --platform=<>` was used then we must
+	// reset context for new stages so that new stages don't inherit unexpected
+	// `--platform` from prior stages.
+	if stage.Builder.Platform != "" || (len(s.stages) > 1 && ((s.systemContext.ArchitectureChoice == "" && s.systemContext.VariantChoice == "" && s.systemContext.OSChoice == "") || (s.executor.skipUnusedStages == types.OptionalBoolFalse))) {
```
Comment (nalind) [inline imagebuildah/stage_executor.go:973]: In the current revision, `s.executor.skipUnusedStages` doesn't appear to matter and can likely be dropped, as the current revision ensures that OSChoice/ArchitectureChoice/VariantChoice is explicitly specified for every stage of a multi-stage build (earlier revisions only did so for the second and later stages), either to values given in the `--platform` flag for the FROM instruction, values passed in through our API, or the current system's defaults.

**Review by nalind — COMMENTED** [review]
The conditional can probably be simplified.  Not sure what one of the `buildah build` runs in one of the tests is testing.

File: tests/bud.bats (hunk 7 above)
Comment (flouthoc) [reply tests/bud.bats]: This part is better covered in this issue https://github.com/containers/buildah/issues/6104 and I will add a different test once I am able to verify if issue exists.

File: imagebuildah/stage_executor.go (hunk 8 above)
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: I removed this.

File: tests/bud/multiarch/Containerfile.unused-stage [hunk 9]
```diff
@@ -0,0 +1,3 @@
+FROM busybox
+RUN arch
+FROM --platform=linux/$FOREIGNARCH busybox AS foreign
```
Comment (nalind) [inline tests/bud/multiarch/Containerfile.unused-stage:3]: This file is no longer being used.

**Review by nalind — COMMENTED** [review]
Drop the unused file, then this should be ready to merge.

File: tests/bud/multiarch/Containerfile.unused-stage (hunk 9 above)
Comment (flouthoc) [reply tests/bud/multiarch/Containerfile.unused-stage]: I have removed this.

**Comment by flouthoc** [comment]
@TomSweeneyRedHat @rhatdan @Honny1 PTAL

**Review by Honny1 — APPROVED** [review]
LGTM
 
/approve
